src/platforms/bool.rkt

```diff
@@ -29,29 +29,36 @@
-
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;; rules ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-
-(define-ruleset* bool-reduce
-                 (bools simplify fp-safe)
-                 #:type ([a bool] [b bool])
-                 [not-true (not (TRUE)) (FALSE)]
-                 [not-false (not (FALSE)) (TRUE)]
-                 [not-not (not (not a)) a]
-                 [not-and (not (and a b)) (or (not a) (not b))]
-                 [not-or (not (or a b)) (and (not a) (not b))]
-                 [and-true-l (and (TRUE) a) a]
-                 [and-true-r (and a (TRUE)) a]
-                 [and-false-l (and (FALSE) a) (FALSE)]
-                 [and-false-r (and a (FALSE)) (FALSE)]
-                 [and-same (and a a) a]
-                 [or-true-l (or (TRUE) a) (TRUE)]
-                 [or-true-r (or a (TRUE)) (TRUE)]
-                 [or-false-l (or (FALSE) a) a]
-                 [or-false-r (or a (FALSE)) a]
-                 [or-same (or a a) a])
+(define-operator-impl (not [x : bool])
+                      bool
+                      #:spec (not x)
+                      #:fpcore (! (not x))
+                      #:fl not
+                      #:identities
+                      ([not-true (not (TRUE)) (FALSE)] [not-false (not (FALSE)) (TRUE)]
+                                                       [not-not (not (not a)) a]
+                                                       [not-and (not (and a b)) (or (not a) (not b))]
+                                                       [not-or (not (or a b)) (and (not a) (not b))]
+                                                       [not-lt (not (< x y)) (>= x y)]
+                                                       [not-gt (not (> x y)) (<= x y)]
+                                                       [not-lte (not (<= x y)) (> x y)]
+                                                       [not-gte (not (>= x y)) (< x y)]))
+
+(define-operator-impl (and [x : bool] [y : bool])
+                      bool
+                      #:spec (and x y)
+                      #:fl and-fn
+                      #:identities
+                      ([and-true-l (and (TRUE) a) a] [and-true-r (and a (TRUE)) a]
+                                                     [and-false-l (and (FALSE) a) (FALSE)]
+                                                     [and-false-r (and a (FALSE)) (FALSE)]
+                                                     [and-same (and a a) a]))
+
+(define-operator-impl (or [x : bool] [y : bool])
+                      bool
+                      #:spec (or x y)
+                      #:fl or-fn
+                      #:identities ([or-true-l (or (TRUE) a) (TRUE)] [or-true-r (or a (TRUE)) (TRUE)]
+                                                                     [or-false-l (or (FALSE) a) a]
+                                                                     [or-false-r (or a (FALSE)) a]
+                                                                     [or-same (or a a) a]))
```

Comment:
Should we just assert that any rules over booleans are fp-safe?

Reply:
Theoretically, you can have different representations that implement bool. But I think in practice this won't happen.